File: question/leetcode/leetcode0152.Maximum-Product-Subarray/leetcode0152.go
package leetcode

func maxProduct(nums []int) int {
	if len(nums) == 0 {
		return 0
	}

	maxProd, minProd, result := nums[0], nums[0], nums[0]
	for i := 1; i < len(nums); i++ {
		// 当前值可能使之前的最小乘积变为最大乘积，因此需要暂存当前的最大乘积
		tempMax := maxProd
		//有负号的原因，最大乘积 可能是  最大乘积* 当前值， 也可能是最小乘积*当前值， 也可以就是当前值
		maxProd = max(max(tempMax*nums[i], minProd*nums[i]), nums[i])
		//同理，最小乘积可能是 当前位置的最大乘积 * 当前值， 也可能是最小乘积*当前值， 也可能是当前值
		minProd = min(min(tempMax*nums[i], minProd*nums[i]), nums[i])
		result = max(result, maxProd)
	}

	return result
}

File: question/leetcode/leetcode0152.Maximum-Product-Subarray/leetcode0152.ts
function maxProduct(nums: number[]): number {
	let n = nums.length;
	if (n === 0) {
		return 0
	}
	let maxVal = nums[0], minVal = nums[0], ans = nums[0]
	for (let i = 1; i < n; i++) {
		let tempMaxVal = maxVal;

		maxVal = Math.max(Math.max(tempMaxVal * nums[i], minVal * nums[i]), nums[i]);
		minVal = Math.min(Math.min(tempMaxVal * nums[i], minVal * nums[i]), nums[i]);
		ans = Math.max(ans, maxVal)
	}

	return ans;
};